### PR TITLE
[BEAM-53] Pub/sub client with gRPC implementation

### PR DESCRIPTION
**Pull request by mshields822**
Support pub/sub via gRPC.
Will later use for pure Java pub/sub source/sink.

### CODE REVIEW

**Comment by mshields822** [comment]
R: @kennknowles

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ * limitations under the License.
+ */
+
+package com.google.cloud.dataflow.sdk.io;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * A helper interface for talking to pub/sub via an underlying transport.
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: I think the official name is "Pubsub", no slash, capital P, lowercase 's'.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  /**
+   * Request the next batch of up to {@code batchSize} messages from {@code subscription}.
+   * Return the received messages, or empty collection if none were available. Does not
+   * wait for messages to arrive. Returned messages will record heir request time
+   * as {@code requestTimeMsSinceEpoch}.
+   *
+   * @throws IOException
+   */
+  Collection<IncomingMessage> pull(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: It would seem pretty normal to have explicit `Subscription` class rather than passing the string name as a key each time. Is there a reason this cannot be done?

For example, if `#pull` did not take a subscription string, but pulled all available messages from all subscriptions in bulk, then I would understand.

Related - publishing and subscribing seem to be two separable activities for pub/sub but this client seems like roughly the (disjoint) union of the two.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+  @Nullable
+  private ManagedChannel publisherChannel;
+
+  /**
+   * Credentials determined from options and environment.
+   */
+  private final GoogleCredentials credentials;
+
+  /**
+   * Label to use for custom timestamps, or {@literal null} if should use pub/sub publish time
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: `{@code null}`

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+      @Nullable String timestampLabel, @Nullable String idLabel,
+      ManagedChannel publisherChannel, GoogleCredentials credentials) {
+    this.timestampLabel = timestampLabel;
+    this.idLabel = idLabel;
+    this.publisherChannel = publisherChannel;
+    this.credentials = credentials;
+  }
+
+  /**
+   * Construct a new pub/sub grpc client. It should be closed via {@link #close} in order
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: As long as you are mentioning `#close`, mention try-with-resources / `Autocloseable`.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+  @Override
+  public int publish(String topic, Iterable<OutgoingMessage> outgoingMessages) throws IOException {
+    PublishRequest.Builder request = PublishRequest.newBuilder()
+                                                   .setTopic(topic);
+    for (OutgoingMessage outgoingMessage : outgoingMessages) {
+      PubsubMessage.Builder message =
+          PubsubMessage.newBuilder()
+                       .setData(ByteString.copyFrom(outgoingMessage.elementBytes));
+
+      if (timestampLabel != null) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: I know the sink is in another PR, but it does seem like it would be better to keep the `timestampLabel` stuff in the sink.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+      Map<String, String> attributes = pubsubMessage.getAttributes();
+
+      // Payload.
+      byte[] elementBytes = pubsubMessage.getData().toByteArray();
+
+      // Timestamp.
+      // Start with pub/sub processing time.
+      Timestamp timestampProto = pubsubMessage.getPublishTime();
+      long timestampMsSinceEpoch = timestampProto.getSeconds() + timestampProto.getNanos() / 1000L;
+      if (timestampLabel != null && attributes != null) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: For reading from Pubsub in Beam, `timestampLabel` is really a part of the `PubsubIO` class rather than the thick pubsub client itself. It doesn't seem fundamental to pull it out here, since it is just stuck on the message and returned.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/UnboundedSource.java [hunk 7]
```diff
@@ -110,8 +110,21 @@ public boolean requiresDeduping() {
      * <p>For example, this could be sending acknowledgement requests to an external
      * data source such as Pub/Sub.
      *
-     * <p>This may be called from any thread, potentially at the same time as calls to the
-     * {@code UnboundedReader} that created it.
+     * <p>Note that:
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/UnboundedSource.java]: Doc updates are nice, but unrelated. Break into separate commit?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java (hunk 1 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java (hunk 2 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: This class a) captures common bookkeeping between unbounded pubsub source/sink and benchmarking code in a way which avoids making any grpc or protoc details public b) provides a hook for (as yet unwritten) mocks. Feels like scope creep for it to go any further.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 3 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 4 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 5 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: In both the source and the sink I found I needed my own message representation. When it came to making this impl public I was forced to avoid any PubsubMessages in the API. By dealing with timestampLabel (and idLabel) here I get to kill two birds as it were: have the intermediate representation I need and solve the API problem.

Remember, every message goes through three representations (T, OutgoingMessage, PubsubMessage). That's already pretty unfortunate.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 6 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: see above.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/UnboundedSource.java (hunk 7 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/UnboundedSource.java]: I'll move to the unbounded source commit. I added them based on my experience implementing that.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: Given that this is blocking other work, I'm OK to wait on developing it into a richer client. What we really need is a thick Pubsub client, which I agree is a slightly larger scope that your immediate goals. No need to do everything up front.

An intermediate approach that leverages the type system to prevent silly errors and improves readability: a simple rich string wrapper to make it explicitly type-incompatible with `String`. Should be ~4 lines using `AutoValue` or ~10 without it. (What I mean is the Java version of the Haskell `newtype Subscription = Subscription String`, if that means anything to you.)

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+    this.timestampLabel = timestampLabel;
+    this.idLabel = idLabel;
+    this.publisherChannel = publisherChannel;
+    this.credentials = credentials;
+  }
+
+  /**
+   * Construct a new Pubsub grpc client. It should be closed via {@link #close} in order
+   * to ensure tidy cleanup of underlying netty resources. (Or use the try-with-resources
+   * construct since this class is {@link AutoCloseable}. If non-{@literal null}, use
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: close paren

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 5 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: I see what you mean. Agreed that it saves some API surface to do it this way.

**Comment by kennknowles** [comment]
Looks like checkstyle dislikes your copyright headers.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java (hunk 2 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: Done. Decided not to add dep on com.google.auto.value:auto-value though it is certainly tempting.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java (hunk 8 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java]: Done.

**Comment by mshields822** [comment]
Could do with a squash but this addresses your comments and supports some quirks needed later by the dataflow runner.

**Comment by mshields822** [comment]
Looking at the misc support code in PubsubIO thinks would be much clearer if we also had a PubsubApiaryClient and all the regex and other misc handling in PubsubIO was moved into PubsubClient.

BUT I'd like to make steady progress, so how about we take that as a TODO?

**Comment by kennknowles** [comment]
LGTM. Merging.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+import com.google.api.client.repackaged.com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collection;
+
+/**
+ * A helper interface for talking to Pubsub via an underlying transport.
+ */
+public interface PubsubClient extends AutoCloseable {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: It occurs to me that this PR would really benefit from a follow-up with an in-memory testing fake, with a test suite that can be applied to both it and the gRPC implementation.

**Comment by davorbonaci** [comment]
I think this PR might not have received the attention it deserves:
* New dependencies; what's the impact on shading and leakages? Impact on dependency tree? What are transitive dependencies?
* New public APIs in the `io` package? This sounds something that better fits `util`.
* No tests.
* Importing repackaged Preconditions?
* Etc.

@mshields822, can I ask you to work with @dhalperi to get these things fixed up?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.dataflow.sdk.io;
+
+import com.google.api.client.repackaged.com.google.common.base.Preconditions;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java]: Whoops

**Comment by kennknowles** [comment]
Package-private would be better than `util/`, and it does seem possible for now at least.

**Comment by mshields822** [comment]
No worries:
 1. Flink can't handle gRPC without special magic (openSSL dep problems), so that's the last straw for me with gRCP. I'm going to add a PubsubApiaryClient and see what I can do to bring the gRPC deps back under control.
 2. Only reason any of this is public is so the Nexmark driver can piggy back on it for all its pubsub create/list/delete stuff. Davor suggested util/ is a happy place for this sort of code, will follow that.
 3. Repackaged slipped through an auto-complete. Sorry about that.
 4. Hard to unit test this and I didn't want to go to town on integration or service tests, especially since it is much easier to test at the pipeline level than the raw client level.
Stand by.

**Comment by mshields822** [comment]
Also, the extra deps (beyond grpc-pubsub-v1) are all needed by the gRPC client. They are also included by the grpc-pubsub-v1 dep, but our transitive dependency check requires them to be included directly. You can find the same deps inside the bigtable-client-core pom. So the only new dep here is on grpc-pubsub-v1 itself.

**Comment by mshields822** [comment]
This work continues in
https://github.com/apache/incubator-beam/pull/213
